63.CUDAInterop/main.cpp

```diff
@@ -0,0 +1,411 @@
+		smart_refctd_ptr<CVulkanConnection> api;
+		{
+			// You generally want to default initialize any parameter structs
+			IAPIConnection::SFeatures apiFeaturesToEnable = {};
+			// generally you want to make your life easier during development
+			apiFeaturesToEnable.validations = true;
+			apiFeaturesToEnable.synchronizationValidation = true;
+			// want to make sure we have this so we can name resources for vieweing in RenderDoc captures
+			apiFeaturesToEnable.debugUtils = true;
+			// create our Vulkan instance
+			if (!(api = CVulkanConnection::create(smart_refctd_ptr(m_system), 0, _NBL_APP_NAME_, smart_refctd_ptr(base_t::m_logger), apiFeaturesToEnable)))
+				return logFail("Failed to crate an IAPIConnection!");
+		}
+
+		// We won't go deep into performing physical device selection in this example, we'll take any device with a compute queue.
+		// Nabla has its own set of required baseline Vulkan features anyway, it won't report any device that doesn't meet them.
+		IPhysicalDevice* physDev = nullptr;
+		ILogicalDevice::SCreationParams params = {};
+		// we will only deal with a single queue in this example
+		params.queueParamsCount = 1;
+		params.queueParams[0].count = 1;
+		params.featuresToEnable;
+		for (auto physDevIt = api->getPhysicalDevices().begin(); physDevIt != api->getPhysicalDevices().end(); physDevIt++)
+		{
+			const auto familyProps = (*physDevIt)->getQueueFamilyProperties();
+			// this is the only "complicated" part, we want to create a queue that supports compute pipelines
+			for (auto i = 0; i < familyProps.size(); i++)
+				if (familyProps[i].queueFlags.hasFlags(IQueue::FAMILY_FLAGS::COMPUTE_BIT))
+				{
+					physDev = *physDevIt;
+					params.queueParams[0].familyIndex = i;
+					break;
+				}
+		}
+		if (!physDev)
+			return logFail("Failed to find any Physical Devices with Compute capable Queue Families!");
+
+		{
+			auto& limits = physDev->getLimits();
+			if (!limits.externalMemoryWin32 || !limits.externalFenceWin32 || !limits.externalSemaphoreWin32)
+				return logFail("Physical device does not support the required extensions");
+			
+			cudaHandler = CCUDAHandler::create(system.get(), smart_refctd_ptr<ILogger>(m_logger));
+			assert(cudaHandler);
+			cudaDevice = cudaHandler->createDevice(smart_refctd_ptr_dynamic_cast<CVulkanConnection>(api), physDev);
+		}
+
+		// logical devices need to be created form physical devices which will actually let us create vulkan objects and use the physical device
+		logicalDevice = physDev->createLogicalDevice(std::move(params));
+		if (!logicalDevice)
+			return logFail("Failed to create a Logical Device!");
+		
+		queue = logicalDevice->getQueue(params.queueParams[0].familyIndex, 0);
+		
+		createResources();
+
+		smart_refctd_ptr<ICPUBuffer> ptx;
+		{
+			ISystem::future_t<smart_refctd_ptr<IFile>> fut;
+			m_system->createFile(fut, "../vectorAdd_kernel.cu", IFileBase::ECF_READ);
```

Comment:
please slap the `.cu` file to a builtin resource archive and use it like ex 03 does

Reply:
this is why you should inherit from `MonoAssetManagerAndBuiltinResourceApplication`